ath79: add support for Fortinet FAP-221-C

bajansen opened the pull request:
FCC ID: TVE-121402

Fortinet FAP-221-C is an indoor access point with 1gb ethernet port, dual-band wireless, internal antenna plates, and 802.3at PoE+.

Hardware and board design are from Senao. The device appears very similar to the EnGenius EAP1200H, albeit with double the flash and RAM.

**Specifications:**

  - QCA9557 SOC
  - QCA9882 WLAN PCI card, 5 GHz, 2x2, 26dBm
  - AR8035-A PHY RGMII GbE with PoE+ IN
  - 40 MHz clock
  - 32 MB FLASH FL256SAIFR0
  - 2x 128 MB RAM NT5TU64M16HG
  - UART populated
  - 4 internal antenna plates
  - 5 LEDs, 1 button (power, 'warning', eth0, wifi1, wifi2) (reset)

  Amber LAN LED appears hardwired to ethernet port. Power LED is green
only. Other LEDs are amber/green.

**MAC addresses:**

1 MAC Address in flash at end of uboot
ASCII encoded, no delimiters
Labeled as "MAC Address" on case

**Serial Access:**

Pinout: (arrow) VCC GND RX TX

Pins are populated with a header and traces not blocked. Bootloader is set to 9600 baud, 8 data, 1 stop.

**Console Access:**

Bootloader:

Interrupt boot with Ctrl+C
Press "k" and enter password "1"
OR
Hold reset button for 5 sec during power on
Interrupt the TFTP transfer with Ctrl+C

to print commands available, enter "help"

OEM:

default username is "admin", password blank
telnet is available at default address 192.168.1.2 serial is available with baud 9600

to print commands available, enter "help"
or tab-tab (busybox list of commands)

**Installation:**

Use factory.bin with OEM upgrade procedures
OR
Use initramfs.bin with uboot TFTP commands.
Then perform a sysupgrade with sysupgrade.bin

**TFTP Recovery:**

Using serial console, load initramfs.bin using TFTP to boot openwrt without touching the flash.

**Return to OEM:**

The best way to return to OEM firmware
is to have a copy of the MTD partitions
before flashing Openwrt.

Backup copies should be made of partitions
"fwconcat0", "loader", and "fwconcat1"
which together is the same flash range
as OEM's "rootfs" and "uimage"
by loading an initramfs.bin
and using LuCI to download the mtdblocks.

It is also possible to extract from the
OEM firmware upgrade image by splitting it up
in parts of lengths that correspond
to the partitions in openwrt
and write them to flash,
after gzip decompression.

After writing to the firmware partitions,
erase the "reserved" partition and reboot.

Signed-off-by: Bernardus Jansen <bernardus@bajansen.nl>

bajansen commented:
This work is heavily based on the commits adding support for the Fortinet FAP-221-B and the EnGenius EAP1200H. Only when getting my fork ready for merging did I come across the seemingly abandoned pull request #11621 containing similar work. Though this pull request was created largely independently, #11621 did inspire the GPIO config included in this PR.

Leo-PL commented on target/linux/ath79/dts/qca9557_fortinet_fap-221-c.dts:
```diff
@@ -0,0 +1,232 @@
+		};
+
+		wifi1_green {
+			label = "green:wifi1";
```
You can use color/function here too, use `color` as usual, and `"LED_FUNCTION_WIFI_2GHZ"` as function. For 5GHz radio, there is `"LED_FUNCTION_WIFI_5GHZ"`. For others - if you can't find a match, `function` property is a string.

Leo-PL commented on target/linux/ath79/dts/qca9557_fortinet_fap-221-c.dts:
```diff
@@ -0,0 +1,232 @@
+		};
+
+		lan_green {
+			label = "green:lan";
```
You could use LED_FUNCTION_ID_LAN here.

Leo-PL commented on target/linux/ath79/dts/qca9557_fortinet_fap-221-c.dts:
```diff
@@ -0,0 +1,232 @@
+		};
+
+		warning_green {
+			label = "green:warning";
```
And "warning", I believe is the only case, when using a custom function would be necessary - but please check the definitions list in kernel - maybe something matches.

Leo-PL reviewed (changes requested):
Ah, yes. Fortinet, never change.
Looks good to me, save for minor adjustments needed for LEDs to match the new convention, but these should be straightforward.
Kudos for very detailed device description!

bajansen commented:
I did add the correct LED function for LAN, but blindly copied the suggestions for the WIFI functions which turned out to be slightly off (WIFI -> WLAN) and as it appeared pretty cosmetic I didn't get around to rebuilding the kernel so I didn't catch it. Everything should be okay now.

Leo-PL reviewed (changes requested):
`label` property is still left around in some LED nodes. Please take a look here: https://github.com/openwrt/openwrt/pull/17550/files - for example how to deal with custom functions, look for "air" and "dir" leds.

Leo-PL reviewed (approved):
Looking well now.

hauke commented:
Please squash the commits together into one commit.

bajansen committed:
ath79: add support for Fortinet FAP-221-C

FCC ID: TVE-121402

Fortinet FAP-221-C is an indoor access point with 1gb ethernet port,
dual-band wireless, internal antenna plates, and 802.3at PoE+.

Hardware and board design are from Senao. The device appears very
similar to the EnGenius EAP1200H, albeit with double the flash and RAM.

**Specifications:**

  - QCA9557 SOC
  - QCA9882 WLAN PCI card, 5 GHz, 2x2, 26dBm
  - AR8035-A PHY RGMII GbE with PoE+ IN
  - 40 MHz clock
  - 32 MB FLASH FL256SAIFR0
  - 2x 128 MB RAM NT5TU64M16HG
  - UART populated
  - 4 internal antenna plates
  - 5 LEDs, 1 button (power, 'warning', eth0, wifi1, wifi2) (reset)

  Amber LAN LED appears hardwired to ethernet port. Power LED is green
only. Other LEDs are amber/green.

**MAC addresses:**

1 MAC Address in flash at end of uboot
ASCII encoded, no delimiters
Labeled as "MAC Address" on case

**Serial Access:**

Pinout: (arrow) VCC GND RX TX

Pins are populated with a header and traces not blocked.
Bootloader is set to 9600 baud, 8 data, 1 stop.

**Console Access:**

Bootloader:

Interrupt boot with Ctrl+C
Press "k" and enter password "1"
OR
Hold reset button for 5 sec during power on
Interrupt the TFTP transfer with Ctrl+C

to print commands available, enter "help"

OEM:

default username is "admin", password blank
telnet is available at default address 192.168.1.2
serial is available with baud 9600

to print commands available, enter "help"
or tab-tab (busybox list of commands)

**Installation:**

Use factory.bin with OEM upgrade procedures
OR
Use initramfs.bin with uboot TFTP commands.
Then perform a sysupgrade with sysupgrade.bin

**TFTP Recovery:**

Using serial console, load initramfs.bin using TFTP
to boot openwrt without touching the flash.

**Return to OEM:**

The best way to return to OEM firmware
is to have a copy of the MTD partitions
before flashing Openwrt.

Backup copies should be made of partitions
"fwconcat0", "loader", and "fwconcat1"
which together is the same flash range
as OEM's "rootfs" and "uimage"
by loading an initramfs.bin
and using LuCI to download the mtdblocks.

It is also possible to extract from the
OEM firmware upgrade image by splitting it up
in parts of lengths that correspond
to the partitions in openwrt
and write them to flash,
after gzip decompression.

After writing to the firmware partitions,
erase the "reserved" partition and reboot.

Signed-off-by: Bernardus Jansen <bernardus@bajansen.nl>
Link: https://github.com/openwrt/openwrt/pull/18109
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

mcprat commented:
> This work is heavily based on the commits adding support for the Fortinet FAP-221-B and the EnGenius EAP1200H

hmm

I added both of those, why did no one `@` me for review?
the only reason I know is because the stale PR got closed.

@bajansen :

`pad-to 11520k`

this number is wrong

It seems to me you studied the stale PR more than my commit,
otherwise, you would have seen that the last pad-to is larger than `IMAGE_SIZE`.

However, I had commented on this issue in that PR... or was I not clear enough? 
https://github.com/openwrt/openwrt/pull/11621#discussion_r1073098473

This also makes me question whether you even tested a factory.bin or not...
So, you have a different number than the stale PR, but still not the right one...
How did you get the number?

By the way, do you have OEM boot logs to prove the flash layout?
I should have asked the last person about boot logs too...

To be clear, the reason we are padding twice is because the final image must have both:
1. the loader written at the exact offset with respect to the new kernel as the flash layout describes, which is also equal to the offset of the OEM kernel and also equal to the size of the OEM rootfs, and,
2. the exact same size as both OEM kernel and rootfs partitions combined (before compression).

otherwise, the loader will not be written at all...
let me know if you need a deeper explanation...

@hauke is this too confusing?
should we maybe have a separate device variable for the 2 different "image sizes"?

bajansen commented:
> @bajansen :
> 
> `pad-to 11520k`
> 
> this number is wrong
> 
> It seems to me you studied the stale PR more than my commit, otherwise, you would have seen that the last pad-to is larger than `IMAGE_SIZE`.
> 
> However, I had commented on this issue in that PR... or was I not clear enough? [#11621 (comment)](https://github.com/openwrt/openwrt/pull/11621#discussion_r1073098473)
>
> This also makes me question whether you even tested a factory.bin or not... So, you have a different number than the stale PR, but still not the right one... How did you get the number?

I re-ran my calculations, and this is a pretty embarrassing case of hex/decimal mixup.
`fwconcat0` (0x140) + `loader` (0x1) + `fwconcat1` (0x3f) = 0x180 = 384
If my math does check out this time, the correct value should be 24576k.

Though the padding value was apparently incorrect, I did check the factory.bin and successfully used it to flash two FAP-221-C's.

> By the way, do you have OEM boot logs to prove the flash layout? I should have asked the last person about boot logs too...

The OEM boot is not very verbose, but the OEM firmware does allow you to check dmesg:
```
<5>5 cmdlinepart partitions found on MTD device ath-nor0
<5>Creating 5 MTD partitions on "ath-nor0":
<5>0x000000000000-0x000000040000 : "u-boot"
<5>0x000000040000-0x000001440000 : "rootfs"
<5>0x000001440000-0x000001840000 : "uImage"
<5>0x000001840000-0x000001ff0000 : "reserved"
<5>0x000001ff0000-0x000002000000 : "caldata"
```
